src/core/components/pin.js

```diff
@@ -0,0 +1,529 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const CID = require('cids')
+const pinSet = require('./pin-set')
+const KeySet = require('./key-set')
+const promisify = require('promisify-es6')
+const multihashes = require('multihashes')
+const Key = require('interface-datastore').Key
+const _ = require('lodash')
+const each = require('async/each')
+const waterfall = require('async/waterfall')
+const until = require('async/until')
+const once = require('once')
+
+const keyString = multihashes.toB58String
+
+module.exports = function pin (self) {
+  let directPins = new KeySet()
+  let recursivePins = new KeySet()
+  let internalPins = new KeySet()
+
+  // temp - should use /local/pins to be consistent with go repo, but
+  // the pin set in datastore in test/go-ipfs-repo doesn't deserialize
+  const pinDataStoreKey = new Key('/local/pins/js')
+
+  const repo = self._repo
+  const dag = self.dag
+
+  function normalizeHashes (hashes, callback) {
+    // try to accept a variety of hash options including
+    // multihash Buffers, base58 strings, and ipfs path
+    // strings, either individually or as an array
+    if (Buffer.isBuffer(hashes) || typeof hashes.forEach !== 'function') {
+      hashes = [hashes]
+    }
+    const normalized = {
+      hashes: [],
+      update: (multihash, cb) => {
+        try {
+          multihashes.validate(multihash)
+        } catch (err) { return cb(err) }
+
+        normalized.hashes.push(multihash)
+        cb()
+      }
+    }
+    each(hashes, (hash, cb) => {
+      if (typeof hash === 'string') {
+        // example: '/ipfs/QmRootHash/links/by/name'
+        const matched = hash.match(/^(?:\/ipfs\/)?([^/]+(?:\/[^/]+)*)\/?$/)
```

Comment:
would be nice to have utils around this :)

Reply:
I actually wonder if I should be doing this at all here, it seems like there should already be some built-in way to follow paths like this using the existing interface? All this is just for the sake of getting the cid of the thing at the end of the path. I could use `files.get` with the full path but I'm not sure how to get the cid from the result. @diasdavid any suggestion?